Add support for RSA3072 and RSA4096.

vathpela opened the pull request.

vathpela committed:
efikeygen: move knowledge of key size and exponents up a level

We need to be able to generate keys other than just RSA2048 now.  As a
result, we need to have the key generation data determined outside of
generate_keys() itself.

This makes those parameters part of the generate_keys() call, and moves
the default values into efikeygen itself.

Signed-off-by: Peter Jones <pjones@redhat.com>

vathpela committed:
efikeygen: Add support for RSA3072 and RSA4096

This adds a "--algorithm" flag to which you can pass rsa2048, rsa3072,
and rsa4096.

Signed-off-by: Peter Jones <pjones@redhat.com>

Blarse commented:
Hi,

I've noticed that your PR does not take into account the change of the ASN.1 type in bundle_signature(). The type offset in the DER encoded signature depends on the size of the signature.

I propose the following patch:
```diff
--- a/src/efikeygen.c
+++ b/src/efikeygen.c
@@ -141,7 +141,8 @@ bundle_signature(cms_context *cms, SECItem *sigder, SECItem *data,
                errx(1, "could not encode certificate: %s",
                        PORT_ErrorToString(PORT_GetError()));
 
-       sigder->data[sigder->len - 261] = DER_BIT_STRING;
+       //Note: offset is signature size + 5 bytes for DER encoding
+       sigder->data[sigder->len - (signature->len + 5)] = DER_BIT_STRING;
 
        return 0;
 }
```

Other than that this PR look good to me, is there any reason not to merge it yet?

vathpela commented:
Well, it wasn't actually working when I tried it, possibly due to this exact issue.  Hopefully I'll get back to it in a week or two and figure out why.

Blarse committed:
efikeygen: Account for the signature size in bundle_signature()

In ea7a2c41f92d, when bundling the signature, the bitstring type field
is being set manually with a hacky offset.  That offset is only valid
with specific signature types, and so with any signature of a different
size, this is just corrupting data either in the signature or after it.

This change from Egor fixes the egregious hack to manually set the type
so that it computes the location based on the signature length, rather
than hard-coding a value.

Signed-off-by: Peter Jones <pjones@redhat.com>

vathpela commented:
Yep, that change gets me from:
`efikeygen: efikeygen.c:main:1287: could not import signature: security library: invalid arguments.`
to working key generation and signatures that validate.  Thanks!
